packages/venia-ui/lib/targets/venia-ui-intercept.js

```diff
@@ -0,0 +1,127 @@
+    );
+
+/**
+ * Webpack can process loaders best when each one has a unique identity. We
+ * use the filename, since there should only be one of these loaders per
+ * compilation.
+ */
+const ident = __filename;
+
+const name = '@magento/venia-ui';
+
+// Guard condition for whether the loader has already been installed.
+const loaderIsInstalled = mod => mod.loaders.some(l => l.ident === ident);
+
+module.exports = targets => {
+    const builtins = targets.of('@magento/pwa-buildpack');
+
+    builtins.specialFeatures.tap(featuresByModule => {
+        featuresByModule['@magento/venia-ui'] = {
+            cssModules: true,
+            esModules: true,
+            graphqlQueries: true,
+            rootComponents: true,
+            upward: true
+        };
+    });
+
+    builtins.webpackCompiler.tap(compiler =>
+        compiler.hooks.compilation.tap(name, compilation =>
+            compilation.hooks.normalModuleLoader.tap(
+                name,
+                (loaderContext, mod) => {
+                    if (isRCR(mod) && !loaderIsInstalled(mod)) {
+                        const renderers = [];
+                        const api = {
+                            add(renderer) {
+                                renderers.push(renderer);
+                            }
+                        };
+                        targets.own.richContentRenderers.call(api);
+                        mod.loaders.push({
+                            ident: ident,
+                            loader,
+                            options: {
+                                renderers: renderers.reverse()
+                            }
+                        });
+                    }
+                }
+            )
+        )
+    );
+
+    // Dogfood our own richContentRenderer hook to insert the fallback renderer.
+    targets.own.richContentRenderers.tap(api =>
+        api.add({
+            componentName: 'PlainHtmlRenderer',
+            importPath: './plainHtmlRenderer'
+        })
+    );
+
+    /**
+     * Implementation of our `routes` target. When Buildpack runs
+     * `transformModules`, this interceptor will provide a nice API to
+     * consumers of `routes`: instead of specifying the transform file
+     * and the path to the routes component, you can just push route
+     * requests into a neat little array.
+     */
+    builtins.transformModules.tap(addTransform => {
+        addTransform({
+            type: 'babel',
+            fileToTransform:
+                '@magento/venia-ui/lib/components/Routes/routes.js',
+            transformModule:
+                '@magento/venia-ui/lib/targets/BabelRouteInjectionPlugin',
+            options: {
+                routes: targets.own.routes.call([])
+            }
+        });
+    });
+
+    targets.own.routes.tap(routes => [
+        ...routes,
+        {
+            name: 'Cart',
+            pattern: '/cart',
+            exact: true,
+            path: '../CartPage'
```

Comment:
Yeah, this should be better commented. I chose to write the big doc comment in venia-ui-declare.js, but since this is the implementation, it should be more obvious what's going on.

As for the path to the component, I think you're right that it needs to be less confusing. The intent is for all external packages to reference their own components by a fully resolvable path, e.g. `path: '@my/extension/CartPage'`. That doesn't work when you're adding a route directly from the project, though, but you can fix that by resolving directly to the file on disk using `require.resolve`.

This definitely needs another pass. What path convention would make the most sense to you? Should we tweak this so that relative paths are always resolved to the module doing the requesting?

Reply:
> What path convention would make the most sense to you? Should we tweak this so that relative paths are always resolved to the module doing the requesting?

I think absolute path resolution would be preferable.